File: src/visual_signature/adapters/firecrawl_adapter.py
"""Firecrawl acquisition adapter for Visual Signature.

This layer only acquires observable web data. Brand3-owned normalization,
taxonomy, and confidence logic live outside this adapter.
"""

from __future__ import annotations

from dataclasses import is_dataclass
from datetime import datetime
from typing import Any

from src.collectors.web_collector import WebCollector, WebData
from src.config import FIRECRAWL_API_KEY
from src.visual_signature.types import (
    ScreenshotSignal,
    VisualAcquisitionResult,
    VisualAssetCandidate,
    VisualSignatureInput,
)


class FirecrawlVisualSignatureAdapter:
    name = "firecrawl"

    def __init__(self, api_key: str | None = None, web_collector: WebCollector | None = None):
        self.api_key = api_key if api_key is not None else FIRECRAWL_API_KEY
        self.web_collector = web_collector or WebCollector(api_key=self.api_key)

    def acquire(self, input_data: VisualSignatureInput) -> VisualAcquisitionResult:
        acquired_at = datetime.now().isoformat()
        if not self.api_key:
            return VisualAcquisitionResult(
                adapter=self.name,
                requested_url=input_data.website_url,
                final_url=input_data.website_url,
                errors=["FIRECRAWL_API_KEY not set"],
                acquired_at=acquired_at,
            )
        try:
            web_data = self.web_collector.scrape(input_data.website_url)
        except Exception as exc:
            return VisualAcquisitionResult(
                adapter=self.name,
                requested_url=input_data.website_url,
                final_url=input_data.website_url,
                errors=[str(exc)],
                acquired_at=acquired_at,
            )
        return acquisition_from_web_data(web_data, adapter=self.name, acquired_at=acquired_at)


def acquisition_from_web_data(
    web_data: WebData | Any,
    *,
    adapter: str = "existing_web_data",
    acquired_at: str | None = None,
    screenshot_payload: dict[str, Any] | None = None,
) -> VisualAcquisitionResult:
    """Build acquisition data from Brand3's already-collected WebData shape."""
    acquired_at = acquired_at or datetime.now().isoformat()
    screenshot_url = (
        _get(web_data, "screenshot_path")
        or (screenshot_payload or {}).get("screenshot_url")
        or (screenshot_payload or {}).get("url")
    )
    screenshot = (
        ScreenshotSignal(url=str(screenshot_url), source="screenshot")
        if screenshot_url
        else None
    )
    metadata = {
        "title": _get(web_data, "title") or "",
        "description": _get(web_data, "meta_description") or "",
        "canonical_url": _get(web_data, "canonical_url") or "",
        "tech_stack": list(_get(web_data, "tech_stack") or []),
        "browser_status": _get(web_data, "browser_status"),
        "load_time_ms": _get(web_data, "load_time_ms"),
    }
    error = _get(web_data, "error") or ""
    return VisualAcquisitionResult(
        adapter=adapter,
        requested_url=_get(web_data, "url") or "",
        final_url=_get(web_data, "canonical_url") or _get(web_data, "url") or "",
        status_code=_get(web_data, "browser_status"),
        rendered_html=_get(web_data, "html") or "",
        raw_html=_get(web_data, "html") or "",
        markdown=_get(web_data, "markdown_content") or "",
        links=list(_get(web_data, "links") or []),
        images=_normalize_images(_get(web_data, "images") or []),
        screenshot=screenshot,
        metadata=metadata,
        warnings=[],
        errors=[str(error)] if error else [],
        acquired_at=acquired_at,
    )


def _normalize_images(items: list[Any]) -> list[VisualAssetCandidate]:
    candidates: list[VisualAssetCandidate] = []
    for item in items:
        if isinstance(item, str):
            candidates.append(
                VisualAssetCandidate(
                    url=item,
                    source="images",
                    role_hint=_role_hint_from_text(item),
                )
            )
            continue
        if not isinstance(item, dict):
            continue
        url = item.get("url") or item.get("src")
        if not url:
            continue
        candidates.append(
            VisualAssetCandidate(
                url=str(url),
                alt=str(item.get("alt") or "") or None,
                width=_int_or_none(item.get("width")),
                height=_int_or_none(item.get("height")),
                source="images",
                role_hint=_role_hint_from_text(f"{url} {item.get('alt') or ''}"),
            )
        )
    return candidates


def _role_hint_from_text(value: str) -> str:
    lowered = value.lower()
    if any(token in lowered for token in ("logo", "brandmark", "wordmark")):
        return "logo"
    if any(token in lowered for token in ("icon", "favicon")):
        return "icon"
    if any(token in lowered for token in ("illustration", "graphic")):
        return "illustration"
    if any(token in lowered for token in ("background", "hero")):
        return "background"
    return "unknown"


def _int_or_none(value: Any) -> int | None:
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _get(value: Any, field_name: str) -> Any:
    if isinstance(value, dict):
        return value.get(field_name)
    if is_dataclass(value) or hasattr(value, field_name):
        return getattr(value, field_name, None)
    return None

File: src/brand3/visual-signature/adapters/firecrawlAdapter.ts
import type {
  FirecrawlAdapterOptions,
  VisualAcquisitionAdapter,
  VisualAcquisitionResult,
  VisualAssetCandidate,
  VisualSignatureInput,
} from "../types";

type FirecrawlScrapeResponse = {
  success?: boolean;
  data?: {
    markdown?: string;
    html?: string;
    rawHtml?: string;
    links?: unknown[];
    images?: unknown[];
    screenshot?: string;
    metadata?: Record<string, unknown>;
    warning?: string;
  };
  error?: string;
  message?: string;
};

const DEFAULT_ENDPOINT = "https://api.firecrawl.dev/v2/scrape";
const DEFAULT_TIMEOUT_MS = 60_000;

export class FirecrawlVisualSignatureAdapter implements VisualAcquisitionAdapter {
  readonly name = "firecrawl" as const;

  private readonly apiKey?: string;
  private readonly endpoint: string;
  private readonly timeoutMs: number;
  private readonly waitForMs: number;
  private readonly includeScreenshot: boolean;
  private readonly includeImages: boolean;
  private readonly zeroDataRetention: boolean;
  private readonly fetchImpl: typeof fetch;

  constructor(options: FirecrawlAdapterOptions = {}) {
    this.apiKey = options.apiKey ?? readEnv("FIRECRAWL_API_KEY");
    this.endpoint = options.endpoint ?? DEFAULT_ENDPOINT;
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.waitForMs = options.waitForMs ?? 2_000;
    this.includeScreenshot = options.includeScreenshot ?? true;
    this.includeImages = options.includeImages ?? true;
    this.zeroDataRetention = options.zeroDataRetention ?? false;
    this.fetchImpl = options.fetchImpl ?? fetch;
  }

  async acquire(input: VisualSignatureInput): Promise<VisualAcquisitionResult> {
    const acquiredAt = new Date().toISOString();
    if (!this.apiKey) {
      return emptyAcquisition(input, acquiredAt, ["FIRECRAWL_API_KEY not set"]);
    }

    const formats: Array<string | Record<string, unknown>> = [
      "markdown",
      "html",
      "rawHtml",
      "links",
    ];
    if (this.includeImages) {
      formats.push("images");
    }
    if (this.includeScreenshot) {
      formats.push({
        type: "screenshot",
        fullPage: false,
        quality: 80,
        viewport: { width: 1440, height: 1200 },
      });
    }

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs + 5_000);
    try {
      const response = await this.fetchImpl(this.endpoint, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          url: input.websiteUrl,
          formats,
          onlyMainContent: false,
          onlyCleanContent: false,
          removeBase64Images: true,
          blockAds: true,
          waitFor: this.waitForMs,
          timeout: this.timeoutMs,
          zeroDataRetention: this.zeroDataRetention,
        }),
        signal: controller.signal,
      });
      const payload = (await response.json().catch(() => ({}))) as FirecrawlScrapeResponse;
      const data = payload.data ?? {};
      const metadata = data.metadata ?? {};
      const warning = payload.error ?? payload.message ?? data.warning;
      const errors = response.ok && payload.success !== false ? [] : [warning || `Firecrawl request failed: ${response.status}`];
      const warnings = response.ok && warning ? [warning] : [];
      return {
        adapter: this.name,
        requestedUrl: input.websiteUrl,
        finalUrl: stringFrom(metadata.sourceURL) || input.websiteUrl,
        statusCode: numberFrom(metadata.statusCode) ?? response.status,
        renderedHtml: data.html ?? "",
        rawHtml: data.rawHtml ?? "",
        markdown: data.markdown ?? "",
        links: normalizeStringList(data.links),
        images: normalizeImages(data.images),
        screenshot: data.screenshot
          ? {
              url: data.screenshot,
              viewport: { width: 1440, height: 1200, fullPage: false },
              source: "screenshot",
            }
          : undefined,
        metadata,
        warnings,
        errors,
        acquiredAt,
      };
    } catch (error) {
      return emptyAcquisition(input, acquiredAt, [
        error instanceof Error ? error.message : "Firecrawl request failed",
      ]);
    } finally {
      clearTimeout(timeout);
    }
  }
}

function emptyAcquisition(
  input: VisualSignatureInput,
  acquiredAt: string,
  errors: string[],
): VisualAcquisitionResult {
  return {
    adapter: "firecrawl",
    requestedUrl: input.websiteUrl,
    finalUrl: input.websiteUrl,
    links: [],
    images: [],
    metadata: {},
    warnings: [],
    errors,
    acquiredAt,
  };
}

function normalizeImages(value: unknown): VisualAssetCandidate[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value
    .map((item): VisualAssetCandidate | null => {
      if (typeof item === "string") {
        return { url: item, source: "images", roleHint: roleHintFromUrl(item) };
      }
      if (!item || typeof item !== "object") {
        return null;
      }
      const record = item as Record<string, unknown>;
      const url = stringFrom(record.url) || stringFrom(record.src);
      if (!url) {
        return null;
      }
      return {
        url,
        alt: stringFrom(record.alt),
        width: numberFrom(record.width),
        height: numberFrom(record.height),
        source: "images",
        roleHint: roleHintFromUrl(`${url} ${stringFrom(record.alt)}`),
      };
    })
    .filter((item): item is VisualAssetCandidate => Boolean(item));
}

function normalizeStringList(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value.filter((item): item is string => typeof item === "string" && item.trim().length > 0);
}

function roleHintFromUrl(value: string): VisualAssetCandidate["roleHint"] {
  const lower = value.toLowerCase();
  if (lower.includes("logo") || lower.includes("brandmark") || lower.includes("wordmark")) {
    return "logo";
  }
  if (lower.includes("icon") || lower.includes("favicon")) {
    return "icon";
  }
  if (lower.includes("illustration") || lower.includes("graphic")) {
    return "illustration";
  }
  if (lower.includes("background") || lower.includes("hero")) {
    return "background";
  }
  return "unknown";
}

function readEnv(name: string): string | undefined {
  const maybeProcess = globalThis as typeof globalThis & {
    process?: { env?: Record<string, string | undefined> };
  };
  return maybeProcess.process?.env?.[name];
}

function stringFrom(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() ? value.trim() : undefined;
}

function numberFrom(value: unknown): number | undefined {
  if (typeof value === "number" && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === "string" && value.trim() && Number.isFinite(Number(value))) {
    return Number(value);
  }
  return undefined;
}
